### ISSUE TITLE
Flipped look-ahead attention mask for transformer

### ISSUE THREAD

**Issue by aquaktus** [pytorch/pytorch#29546]
## 🐛 Bug

The look-ahead mask generated for the transformer decoder is flipped allowing initial tokens to use future information and later tokens to get less. This is due to an error in the `generate_square_subsequent_mask` function which casts the look-ahead matrix to `float()` too late causing unexpected behaviour.

## To Reproduce

```
>>> import torch.nn as nn
>>> model = nn.Transformer()
>>> model.generate_square_subsequent_mask(4)
 tensor([[0., 0., 0., 0.],
        [-inf, 0., 0., 0.],
        [-inf, -inf, 0., 0.],
        [-inf, -inf, -inf, 0.]])
```

## Expected behavior

The transformer mask generator is meant to create a mask that prevents information from future tokens to affect current predictions, thus preventing them from seeing future tokens.

Expected output:
```
tensor([[0., -inf, -inf, -inf],
        [0., 0., -inf, -inf],
        [0., 0., 0., -inf],
        [0., 0., 0., 0.]])
```

From the [transformer docs](https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html) 
It is not producing the expected behaviour.
```
def generate_square_subsequent_mask(self, sz):
        r"""Generate a square mask for the sequence. The masked positions are filled with float('-inf').
            Unmasked positions are filled with float(0.0).
        """
        mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
        mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
        return mask
```

If we decompose the function.
```
>>> sz = 4
>>> mask = (torch.triu(torch.ones(sz, sz))==1).transpose(0, 1)
>>> mask
tensor([[ True, False, False, False],
        [ True,  True, False, False],
        [ True,  True,  True, False],
        [ True,  True,  True,  True]])
>>> mask.float().masked_fill(mask == 0, float('-inf'))
tensor([[1., 0., 0., 0.],
        [-inf, 1., 0., 0.],
        [-inf, -inf, 1., 0.],
        [-inf, -inf, -inf, 1.]])
>>> mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
tensor([[0., 0., 0., 0.],
        [-inf, 0., 0., 0.],
        [-inf, -inf, 0., 0.],
        [-inf, -inf, -inf, 0.]])
```

Not casting the mask matrix to float earlier is causing unexpected behaviour with masked fill and the boolean matrix referenced here `masked_fill(mask == 0, float('-inf')` .

## Solution

Cast the mask to float earlier so that interactions with masked fill are as expected.
```
>>> sz = 4
>>> mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1).float()
>>> mask = mask.masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
>>> mask
tensor([[0., -inf, -inf, -inf],
        [0., 0., -inf, -inf],
        [0., 0., 0., -inf],
        [0., 0., 0., 0.]])
```

## Environment

PyTorch version: 1.3.0
Is debug build: No
CUDA used to build PyTorch: 10.1.243

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: TITAN RTX
Nvidia driver version: 430.26
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.3

Versions of relevant libraries:
[pip3] numpy==1.17.3
[pip3] pytorch-nlp==0.4.1
[pip3] torch==1.3.0
[pip3] torchtext==0.4.0
[pip3] torchvision==0.4.1
[conda] Could not collect

## Additional context

The result of this bug is a transformer that trains with the wrong attention look-ahead mask. In this [official PyTorch tutorial](https://pytorch.org/tutorials/beginner/transformer_tutorial.html) the problem isn't showcased since at evaluation time only the loss is measured without any inferred tokens being passed back into the model. 

The consequence of this bug is apparent when training a transformer, and seeing that future tokens in the decoder input affect the prediction scores of a previous token. I can provide code showcasing this in a CoLab notebook if necessary.

**Comment by ngimel** [comment]
cc  @zhangguanheng66

**Comment by zhangguanheng66** [comment]
Have you updated to v.1.3.1 or the master branch? The type-promotion bug should be fixed. Feel free to re-open the issue if it's still there.

**Comment by aquaktus** [comment]
Looks like it's fixed :)

**Comment by exfoxz** [comment]
I'm now running into this issue and seeing two different behaviors with 2 different versions of PyTorch:
Code:
```
sz = 3
mask = (torch.triu(torch.ones(sz, sz)) == 1).transpose(0, 1)
mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
```

PyTorch 1.4.0a0+b9b9fd4 (compiled from source):
```
tensor([[0., 0., 0.],
        [-inf, 0., 0.],
        [-inf, -inf, 0.]])
```

PyTorch 1.5.0a0+f9ad552 (compiled from source):
```
tensor([[0., -inf, -inf],
        [0., 0., -inf],
        [0., 0., 0.]])
```

I suppose the later behavior is the correct one while PyTorch 1.4.0 seems to be giving the wrong mask.
